[TexMap] Incorrect damage when 3d transform added dynamically

Scony opened the pull request:
#### faf48e0c615cab82085ae7c74f1da502b81d3998
<pre>
[TexMap] Incorrect damage when 3d transform added dynamically
<a href="https://bugs.webkit.org/show_bug.cgi?id=288471">https://bugs.webkit.org/show_bug.cgi?id=288471</a>

Reviewed by Nikolas Zimmermann.

This change fixes the damage inference logic for TextureMapperLayers
that are newly added or removed.

In case of newly added TextureMapperLayers that have the transform,
the key is not to take the damage from before transform change into account.

In case of layer removals, the area ocuppied by the removed layer needs to
be damaged as a whole as this is most likely what changed visually -
we cannot assume otherwise anyway.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/damage/layer-downsize-with-movement-expected.txt: Added.
* LayoutTests/platform/glib/damage/layer-downsize-with-movement.html: Added.
* LayoutTests/platform/glib/damage/transform-3d-dynamic-re-addition-expected.txt: Added.
* LayoutTests/platform/glib/damage/transform-3d-dynamic-re-addition.html: Added.
* LayoutTests/platform/glib/damage/transform-3d-dynamic-removal-expected.txt: Added.
* LayoutTests/platform/glib/damage/transform-3d-dynamic-removal.html: Added.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::computeTransformsRecursive):
(WebCore::TextureMapperLayer::setDamage):
(WebCore::TextureMapperLayer::collectDamageSelf):
(WebCore::TextureMapperLayer::damageWholeLayerIncludingItsRectFromPreviousFrame):
(WebCore::TextureMapperLayer::removeFromParent):
(WebCore::TextureMapperLayer::setSize):
(WebCore::TextureMapperLayer::damageWholeLayerDueToTransformChange): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::canInferDamage const):

Canonical link: <a href="https://commits.webkit.org/291885@main">https://commits.webkit.org/291885@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eaaad80a91148bca02e70476ff5da81845312459

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93964 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3288 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98972 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44491 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96014 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21980 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71700 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29049 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96966 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10273 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84873 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52039 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9959 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2531 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43807 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80216 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2613 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101012 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21016 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15312 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80708 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21268 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80855 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80069 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20024 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24615 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1978 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14173 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21000 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26178 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20687 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24147 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22428 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -448,6 +448,7 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
         m_receivedDamage = { };
         m_inferredLayerDamage = { };
         m_inferredGlobalDamage = { };
+        m_damageCollectedAtLeastOnce = true;
```
Is there a way we could change this variable name so instead of saying _what_ ("damage has been collected at least once"), it said why or when? 

> This change fixes the damage inference logic in a way that newly added
TextureMapperLayers are not taking the damage from before transform
change into account.

This is the _why_ -> we want to make sure the transform is taken into account. That reason can easily be forgotten after a while, and the variable name won't convey it

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -290,6 +290,7 @@ class TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLayer> {
     Damage m_inferredLayerDamage; // In layer coordinate space.
     Damage m_inferredGlobalDamage; // In global coordinate space.
     FloatRect m_accumulatedOverlapRegionDamage;
+    bool m_wasLayerTransformedBefore { false };
```
You could add : 1 here.

Scony replied:
This member is no more actually

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -466,39 +467,31 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
+
+    // Use the damage information we received from the GraphicsLayer
+    // along with damage information we inferred while performing
+    // layer-level operations such as resizes, transformations etc.
```
```suggestion
    // layer-level operations such as resizes, transformations, etc.
```

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -1308,8 +1304,8 @@ void TextureMapperLayer::setSize(const FloatSize& size)
         // FIXME: Damage only affected area.
-        m_inferredLayerDamage.add(FloatRect(FloatPoint::zero(), m_state.size));
-        m_inferredLayerDamage.add(FloatRect(FloatPoint::zero(), size));
+        m_damageInLayerCoordinateSpace.add(FloatRect(FloatPoint::zero(), m_state.size));
```
re-reading the comment above, something we might do is extracting this logic into a separate method. Something like

```
const auto& layerResizedDamage = getLayerResizedDamage(/* maybe passing in old and new */);
for (const auto& damage : layerResizedDamage)
    m_damageInLayerCoordinateSpace.add(damage);
```

then algorithmically this method (`setSize`) becomes correct, and it's only the new method that needs a `notImplemented()` + the default, more expensive, implementation you suggested above

Scony replied:
I think this will reduce both readability and performance, so I'm not sure

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -543,8 +536,9 @@ void TextureMapperLayer::damageWholeLayerDueToTransformChange(const Transformati
 {
     // When the layer's transform changes, we must not only damage whole layer using new transform,
     // but also using old transform to cover the area not affected by layer anymore.
```
```suggestion
    // When the layer's transform changes, we must not only damage the whole layer using the new transform,
    // but also using the old transform, as it could previously have affected an area the new one doesn't.
```

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -466,39 +467,31 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
         return;
     }
 
+    if (m_contentsLayer) {
+        // Layers with content layer are fully damaged for now.
+        // FIXME: Remove that special case.
+        m_damageInLayerCoordinateSpace.add(targetRect);
+    }
```
this could be another candidate for extracting out and having a `notImplemented`. 

(extracting this out is only useful if it is something somebody could address on their own, rather stand-alone-ish, as it would help having a signature of expected input and output)

Scony replied:
I'd keep it local as this is temporary W/A important in the context of `collectDamageSelf`

spenap reviewed (commented):
Looks good, and I'm happy with `m_wasLayerTransformedBefore`. Re-reading the comments and the logic, I think it can be understood well  :+1:

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -539,12 +534,15 @@ void TextureMapperLayer::collectDamageSelfChildrenFilterAndMask(TextureMapperPai
+
+    // In many cases damaging whole layer in the "new" state is not enough.
+    // When e.g. changing size, transfrom, etc. the layer (or its part) effectively disappears from one place
+    // and re-appears in the other. Therefore the damaging of a layer in the "old" state is required as well.
```
```suggestion
    // In many cases, damaging th whole layer in the "new" state is not enough.
    // When e.g. changing size, transform, etc. the layer (or its parts) effectively disappears from one place
    // and re-appears in another. Therefore the damaging of a layer in the "old" state is required as well.
```

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -444,10 +445,10 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
 {
     ASSERT(m_damagePropagation);
 
+    m_previousLayerRectInGlobalCoordinateSpace = { };
```
```suggestion
    m_previousLayerRectInGlobalCoordinateSpace.reset();
```

this is a personal preference, but calling `reset()` feels more explicit to me. It makes it clear that this is an optional. Otherwise I might expect that this is the same as the other variables below

spenap reviewed (commented):
LGTM. I do prefer your using `std::optional` instead of a boolean, nice one :+1:

nikolaszimmermann reviewed (approved):
Looks great - agree with @spenap.

Scony committed:
[TexMap] Incorrect damage when 3d transform added dynamically
https://bugs.webkit.org/show_bug.cgi?id=288471

Reviewed by Nikolas Zimmermann.

This change fixes the damage inference logic for TextureMapperLayers
that are newly added or removed.

In case of newly added TextureMapperLayers that have the transform,
the key is not to take the damage from before transform change into account.

In case of layer removals, the area ocuppied by the removed layer needs to
be damaged as a whole as this is most likely what changed visually -
we cannot assume otherwise anyway.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/damage/layer-downsize-with-movement-expected.txt: Added.
* LayoutTests/platform/glib/damage/layer-downsize-with-movement.html: Added.
* LayoutTests/platform/glib/damage/transform-3d-dynamic-re-addition-expected.txt: Added.
* LayoutTests/platform/glib/damage/transform-3d-dynamic-re-addition.html: Added.
* LayoutTests/platform/glib/damage/transform-3d-dynamic-removal-expected.txt: Added.
* LayoutTests/platform/glib/damage/transform-3d-dynamic-removal.html: Added.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::computeTransformsRecursive):
(WebCore::TextureMapperLayer::setDamage):
(WebCore::TextureMapperLayer::collectDamageSelf):
(WebCore::TextureMapperLayer::damageWholeLayerIncludingItsRectFromPreviousFrame):
(WebCore::TextureMapperLayer::removeFromParent):
(WebCore::TextureMapperLayer::setSize):
(WebCore::TextureMapperLayer::damageWholeLayerDueToTransformChange): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::canInferDamage const):

Canonical link: https://commits.webkit.org/291885@main
